### PR TITLE
tinyformat: Add format_no_throw

### PR DESCRIPTION
**Pull request by maflcko**
This adds a `tinyformat::format_no_throw` that (on error) returns the format string with the error message. `format_no_throw` is currently only used in logging, but can be used by other modules after this patch.

The fist commit reverts some style-changes that have been made to tinyformat, which is to be treated like a "subtree", so style-fixes should go upstream and are not acceptable in Bitcoin Core.

### CODE REVIEW

**Comment by laanwj** [comment]
Are you intending to use this in other places than logging?

**Comment by MarcoFalke** [comment]
Yes, the gui should probably use it for messages, so that at least the generic error is logged (as opposed to only "A tfm error occured!")

File: src/logging.h [hunk 1]
```diff
@@ -133,14 +133,7 @@ template <typename... Args>
...
     if (LogInstance().Enabled()) {
-        std::string log_msg;
-        try {
-            log_msg = tfm::format(fmt, args...);
-        } catch (tinyformat::format_error& fmterr) {
-            /* Original format string will have newline so don't add one here */
-            log_msg = "Error \"" + std::string(fmterr.what()) + "\" while formatting log message: " + fmt;
-        }
-        LogInstance().LogPrintStr(log_msg);
+        LogInstance().LogPrintStr(tfm::format_no_throw(fmt, args...));
```
Comment (practicalswift) [inline src/logging.h]: I suggest making `LogPrintf` `noexcept`, but before doing so note that the following call chain is triggered here (also before this PR):
* `LogPrintStr` → `LogTimestampStr` → `FormatISO8601DateTime` → `strprintf`

Generally `strprintf` may throw `tinyformat::format_error`.

In the case of `FormatISO8601DateTime` I don't think that is possible since the format string is well formed. However, that is not understood by the compiler which must assume that `FormatISO8601DateTime` can throw.

I suggest handling the (non-reachable) exception explicitly in `FormatISO8601DateTime` and then making that function `noexcept` too. (Should be done also for its companion `FormatISO8601Date`).

So in summary my suggestion is to:
* Handle `tinyformat::format_error` in `FormatISO8601DateTime` and `FormatISO8601Date`
* Add `noexcept` to `FormatISO8601DateTime`, `FormatISO8601Date` and `LogPrintf`

File: src/tinyformat.h [hunk 2]
```diff
@@ -1052,13 +1053,24 @@ TINYFORMAT_FOREACH_ARGNUM(TINYFORMAT_MAKE_FORMAT_FUNCS)
...
+std::string format(const std::string& fmt, const Args&... args)
 {
-    std::ostringstream oss;
-    format(oss, fmt.c_str(), args...);
-    return oss.str();
+    return format(fmt.c_str(), args...);
+}
+
+template <typename... Args>
+std::string format_no_throw(const std::string& fmt, const Args&... args)
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/tinyformat.h]: How can I (or the compiler) be sure that no other exceptions (e.g. `std::out_of_range`) are thrown?

[the remark this answers is not in the document]
Comment (practicalswift) [reply src/tinyformat.h]: There is really no shortcut here AFAIK: reading + reasoning would be required :-)

Do you see any places where `std::out_of_range` could be thrown that is reachable from `format`?

[the remark this answers is not in the document]
Comment (sipa) [reply src/tinyformat.h]: I think the point of adding "noexcept" isn't so much because you know a function won't throw; it's that you know that exceptions thrown anyway can be regarded as fatal errors.

[the remark this answers is not in the document]
Comment (practicalswift) [reply src/tinyformat.h]: @sipa Stroustrup and Sutter:

> [Use noexcept when exiting a function because of a throw is impossible or unacceptable.](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#e12-use-noexcept-when-exiting-a-function-because-of-a-throw-is-impossible-or-unacceptable)

[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/tinyformat.h]: I see, so it is equivalent to `std::terminate`. I think I will leave that for later, as I am not sure that immediate termination is better than offering the gui or daemon to catch the exception

[the remark this answers is not in the document]
Comment (practicalswift) [reply src/tinyformat.h]: I think it is unwise to give the caller the impression that the function cannot throw by calling it `format_no_throw` if we're not willing to back that with a `noexcept` guarantee.

If it is potentially throwing it shouldn't be called `no_throw` :-)

[the remark this answers is not in the document]
Comment (promag) [reply src/tinyformat.h]: Maybe add a comment like `This function wraps format() to return error messages instead of throwing format_error. All other format() exceptions are thrown away.`

[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/tinyformat.h]: Added a comment

**Comment by MarcoFalke** [comment]
We have a linter to check the number of arguments, so I guess this isn't much needed
